File: java/src/main/java/com/leetcode/SevenFour.java
package com.leetcode;

/**
 * @author: John
 * @date: 2020-02-07 11:59
 * @description: 74
 **/
public class SevenFour {
    public static boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0) {
            return false;
        }
        int start = 0, rows = matrix.length, cols = matrix[0].length;
        int end = rows * cols - 1;
        while (start <= end) {
            int mid = (start + end) / 2;
            if (matrix[mid / cols][mid % cols] == target) {
                return true;
            }
            if (matrix[mid / cols][mid % cols] < target) {
                start = mid + 1;
            } else {
                end = mid -1;
            }
        }
        return false;
    }
}

File: leetcodecpp/leetcode/SevenFour.cpp
//
// Created by John on 2020/2/7.
//

#include "SevenFour.h"
#include <vector>

using namespace std;


class Solution {
public:
    static bool searchMatrix(vector<vector<int> > &matrix, int target) {
        int n = matrix.size(); // the size of rows
        int m = matrix[0].size(); // the size of single row
        int l = 0, r = m * n - 1;
        while (l != r) {
            int mid = (l + r -1) >> 1;
            if (matrix[mid / m][mid % m] < target) {
                l = mid + 1;
            } else {
                r = mid;
            }
        }
        return matrix[r / m][r % m] == target;
    }
};
